ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java

```diff
@@ -0,0 +1,333 @@
+      }
+    }
+
+    while (!nodesQueue.isEmpty()) {
+      AnimatedNode nextNode = nodesQueue.poll();
+      if (nextNode.mChildren != null) {
+        for (int i = 0; i < nextNode.mChildren.size(); i++) {
+          AnimatedNode child = nextNode.mChildren.get(i);
+          child.mActiveIncomingNodes++;
+          if (child.mBFSColor != mAnimatedGraphBFSColor) {
+            child.mBFSColor = mAnimatedGraphBFSColor;
+            activeNodesCount++;
+            nodesQueue.add(child);
+          }
+        }
+      }
+    }
+
+    // STEP 2
+    // BFS over the graph of active nodes in topological order -> visit node only when all its
+    // "predecessors" in the graph have already been visited. It is important to visit nodes in that
+    // order as they may often use values of their predecessors in order to calculate "next state"
+    // of their own. We start by determining the starting set of nodes by looking for nodes with
+    // `mActiveIncomingNodes = 0` (those can only be the ones that we start BFS in the previous
+    // step). We store number of visited nodes in this step in `updatedNodesCount`
+
+    mAnimatedGraphBFSColor++;
+    if (mAnimatedGraphBFSColor == AnimatedNode.INITIAL_BFS_COLOR) {
+      // see reasoning for this check a few lines above
+      mAnimatedGraphBFSColor++;
+    }
+
+    // find nodes with zero "incoming nodes", those can be either nodes from `mUpdatedNodes` or
+    // ones connected to active animations
+    for (int i = 0; i < mUpdatedNodes.size(); i++) {
+      AnimatedNode node = mUpdatedNodes.get(i);
+      if (node.mActiveIncomingNodes == 0 && node.mBFSColor != mAnimatedGraphBFSColor) {
+        node.mBFSColor = mAnimatedGraphBFSColor;
+        updatedNodesCount++;
+        nodesQueue.add(node);
+      }
+    }
+    for (int i = 0; i < mActiveAnimations.size(); i++) {
+      AnimationDriver animation = mActiveAnimations.get(i);
+      AnimatedNode valueNode = animation.mAnimatedValue;
+      if (valueNode.mActiveIncomingNodes == 0 && valueNode.mBFSColor != mAnimatedGraphBFSColor) {
+        valueNode.mBFSColor = mAnimatedGraphBFSColor;
+        updatedNodesCount++;
+        nodesQueue.add(valueNode);
+      }
+    }
+
+    // Run main "update" loop
+    while (!nodesQueue.isEmpty()) {
+      AnimatedNode nextNode = nodesQueue.poll();
+      if (nextNode instanceof PropsAnimatedNode) {
+        // Send property updates to native view manager
+        ((PropsAnimatedNode) nextNode).updateView(uiImplementation);
+      }
+      if (nextNode.mChildren != null) {
+        for (int i = 0; i < nextNode.mChildren.size(); i++) {
+          AnimatedNode child = nextNode.mChildren.get(i);
+          child.mActiveIncomingNodes--;
+          if (child.mBFSColor != mAnimatedGraphBFSColor && child.mActiveIncomingNodes == 0) {
+            child.mBFSColor = mAnimatedGraphBFSColor;
+            updatedNodesCount++;
+            nodesQueue.add(child);
+            child.update(nextNode, frameTimeNanos);
```

Comment:
Why don't we just call this when we pop the child node off the queue? Seems like it would be more consistent with how we process the other nodes.

Also, why are we passing in the last parent to be visited to child.update (as nextNode)?